### PR TITLE
Fix #4535: Handle Null and Nothing in completions

### LINKED ISSUE #4535: Unexpected completion content
When I tried to complete null, I accidentally received the collection method:
<pre><font color="#3465A4">scala&gt; </font><font color="#C4A000">null</font>. 
Elements               addString              count                  getChars               lines                  reduceRightOption      stripLineEnd           toMap
Self                   aggregate              deep                   getClass               linesIterator          regionMatches          stripMargin            toSeq
WithFilter             andThen                diff                   getStackTraceString    linesWithSeparators    replace                stripPrefix            toSet
!=                     apply                  distinct               groupBy                map                    replaceAll             stripSuffix            toShort
##                     applyOrElse            drop                   grouped                matches                replaceAllLiterally    subSequence            toStream
%                      array                  dropRight              hasDefiniteSize        max                    replaceFirst           substring              toString
&amp;                      asInstanceOf           dropWhile              hashCode               maxBy                  repr                   sum                    toTraversable
&amp;&amp;                     canEqual               elemManifest           head                   min                    reverse                synchronized           toUpperCase
*                      capitalize             elemTag                headOption             minBy                  reverseIterator        tail                   toVector
+                      charAt                 endsWith               indexOf                mkString               reverseMap             tails                  transform
++                     chars                  ensuring               indexOfSlice           ne                     reversed               take                   transpose
++:                    clone                  eq                     indexWhere             newBuilder             runWith                takeRight              trim
+:                     codePointAt            equals                 indices                nonEmpty               sameElements           takeWhile              unary_!
-                      codePointBefore        equalsIgnoreCase       init                   notify                 scan                   thisCollection         unary_+
-&gt;                     codePointCount         exists                 inits                  notifyAll              scanLeft               to                     unary_-
/                      codePoints             filter                 intern                 offsetByCodePoints     scanRight              toArray                unary_~
/:                     collect                filterNot              intersect              orElse                 segmentLength          toBoolean              union
:+                     collectFirst           finalize               invert                 padTo                  self                   toBuffer               unzip
:\                     combinations           find                   isDefinedAt            par                    seq                    toByte                 unzip3
&lt;                      companion              flatMap                isEmpty                parCombiner            size                   toChar                 update
&lt;&lt;                     compare                flatten                isInstanceOf           partition              sizeHintIfCheap        toCharArray            updated
&lt;=                     compareTo              fold                   isTraversableAgain     patch                  slice                  toCollection           view
&lt;init&gt;                 compareToIgnoreCase    foldLeft               iterator               permutations           sliding                toDouble               wait
==                     compose                foldRight              last                   prefixLength           sortBy                 toFloat                withFilter
&gt;                      concat                 forall                 lastIndexOf            product                sortWith               toIndexedSeq           zip
&gt;=                     contains               foreach                lastIndexOfSlice       r                      sorted                 toInt                  zipAll
&gt;&gt;                     containsSlice          format                 lastIndexWhere         reduce                 span                   toIterable             zipWithIndex
&gt;&gt;&gt;                    contentEquals          formatLocal            lastOption             reduceLeft             split                  toIterator             zipped
^                      copyToArray            formatted              length                 reduceLeftOption       splitAt                toList                 |
__arrayOfChars         copyToBuffer           genericBuilder         lengthCompare          reduceOption           startsWith             toLong                 ||
__sequenceOfChars      corresponds            getBytes               lift                   reduceRight            stringPrefix           toLowerCase
</pre>
But I can't call them:
<pre><font color="#3465A4">scala&gt; </font><font color="#C4A000">null</font>.isEmpty 
<font color="#CC0000">1 |</font><font color="#C4A000">null</font>.isEmpty
  |<font color="#CC0000">^^^^^^^^^^^^</font>
  |value `isEmpty` is not a member of Null

</pre>
And the `<init>` method also appears in the completion content:
<pre><font color="#3465A4">scala&gt; </font><font color="#CC0000">0</font>. 
ResultWithoutStep         &lt;init&gt;                    compareTo                 isInfinite                max                       toBinaryString            unary_+
!=                        ==                        doubleValue               isInfinity                min                       toByte                    unary_-
##                        &gt;                         ensuring                  isInstanceOf              ne                        toChar                    unary_~
%                         &gt;=                        eq                        isNaN                     notify                    toDegrees                 underlying
&amp;                         &gt;&gt;                        equals                    isNegInfinity             notifyAll                 toDouble                  unifiedPrimitiveEquals
*                         &gt;&gt;&gt;                       finalize                  isPosInfinity             num                       toFloat                   unifiedPrimitiveHashcode
+                         ^                         floatValue                isValidByte               ord                       toHexString               until
-                         abs                       floor                     isValidChar               round                     toInt                     wait
-&gt;                        asInstanceOf              formatted                 isValidInt                self                      toLong                    |
/                         byteValue                 getClass                  isValidLong               shortValue                toOctalString
&lt;                         ceil                      hashCode                  isValidShort              signum                    toRadians
&lt;&lt;                        clone                     intValue                  isWhole                   synchronized              toShort
&lt;=                        compare                   integralNum               longValue                 to                        toString
<font color="#3465A4">scala&gt; </font>(). 
!=            +             &lt;init&gt;        asInstanceOf  equals        getClass      isInstanceOf
##            -&gt;            ==            ensuring      formatted     hashCode      toString</pre>

### PR DESCRIPTION
**Pull request by nicolasstucki**
We special completion on qualifiers of type `Nothing` to not return the method on `Any`.
Also special case completion on qualifiers of type `Null` to not return completion that involves implicit conversions.

Add missing Artifact to remove `$isInstanceOf` completion and fix missing decoding for `$uXXXX` (`$u2192` to `→`).

### CODE REVIEW

File: compiler/src/dotty/tools/dotc/interactive/Completion.scala [hunk 1]
```diff
@@ -207,11 +209,17 @@ object Completion {
...
+        ()
+      else if (qual.tpe =:= defn.NullType)
+        addAccessibleMembers(defn.AnyRefType)
+      else {
+        addAccessibleMembers(qual.tpe)
+        if (!mode.is(Mode.Import)) {
+          // Implicit conversions do not kick in when importing
+          implicitConversionTargets(qual)(ctx.fresh.setExploreTyperState())
+            .foreach(addAccessibleMembers)
+        }
```
Comment (Blaisorblade) [inline compiler/src/dotty/tools/dotc/interactive/Completion.scala]: This makes sense, except is seems a bit hacky?

File: compiler/test/dotty/tools/repl/TabcompleteTests.scala [hunk 2]
```diff
@@ -97,4 +97,25 @@ class TabcompleteTests extends ReplTest {
...
+    assertEquals(
+      List("!=", "##", "+", "->", "==", "asInstanceOf", "clone", "ensuring", "eq", "equals", "finalize", "formatted",
+          "getClass", "hashCode", "isInstanceOf", "ne", "notify", "notifyAll", "synchronized", "toString", "wait", "→"),
+      comp.distinct.sorted)
+  }
+
+  @Test def `???` = fromInitialState { implicit s =>
+    val comp = tabComplete("???.")
+    assertEquals(Nil, comp)
+  }
```
Comment (Blaisorblade) [inline compiler/test/dotty/tools/repl/TabcompleteTests.scala]: 👍

File: compiler/src/dotty/tools/dotc/interactive/Completion.scala [hunk 3]
```diff
@@ -150,7 +150,9 @@ object Completion {
       nameToSymbols.map { case (name, symbols) =>
         val typesFirst = symbols.sortWith((s1, s2) => s1.isType && !s2.isType)
         val desc = description(typesFirst)
-        Completion(name.toString, desc, typesFirst)
+        val strName = name.toString
+        val label = if (strName == "$u2192") "→" else strName // TODO fix name.decode and use it
+        Completion(label, desc, typesFirst)
```
Comment (Blaisorblade) [inline compiler/src/dotty/tools/dotc/interactive/Completion.scala]: This *looks* unrelated and untested, and more fragile than the rest? If it's worth the technical debt, okay...

Comment (nicolasstucki) [reply compiler/src/dotty/tools/dotc/interactive/Completion.scala]: I will open a separate issue for this when we merge it

File: compiler/src/dotty/tools/dotc/interactive/Completion.scala (hunk 1 above)
Comment (nicolasstucki) [reply compiler/src/dotty/tools/dotc/interactive/Completion.scala]: Why hacky? Do you have another idea of how to handle bottom types correctly?

File: compiler/src/dotty/tools/dotc/interactive/Completion.scala [hunk 4]
```diff
@@ -207,11 +209,17 @@ object Completion {
...
      */
     def addMemberCompletions(qual: Tree)(implicit ctx: Context): Unit = {
-      addAccessibleMembers(qual.tpe)
-      if (!mode.is(Mode.Import)) {
-        // Implicit conversions do not kick in when importing
-        implicitConversionTargets(qual)(ctx.fresh.setExploreTyperState())
-          .foreach(addAccessibleMembers)
+      if (qual.tpe =:= defn.NothingType)
+        ()
+      else if (qual.tpe =:= defn.NullType)
```
Comment (smarter) [inline compiler/src/dotty/tools/dotc/interactive/Completion.scala]: I don't understand, what other methods than those on AnyRef does completion find on null ? Where do they come from ?

File: compiler/src/dotty/tools/dotc/interactive/Completion.scala [hunk 5]
```diff
@@ -150,7 +150,9 @@ object Completion {
       nameToSymbols.map { case (name, symbols) =>
         val typesFirst = symbols.sortWith((s1, s2) => s1.isType && !s2.isType)
         val desc = description(typesFirst)
-        Completion(name.toString, desc, typesFirst)
+        val strName = name.toString
+        val label = if (strName == "$u2192") "→" else strName // TODO fix name.decode and use it
```
Comment (smarter) [inline compiler/src/dotty/tools/dotc/interactive/Completion.scala]: Please don't add hacks like that.

File: compiler/src/dotty/tools/dotc/interactive/Completion.scala (hunk 4 above)
Comment (odersky) [reply compiler/src/dotty/tools/dotc/interactive/Completion.scala]: Instead if `=:=` it's more idiomatic to use `isRef`. I.e. `qual.tpe.isRef(defn.NothingClass)`. (It's also faster).

File: compiler/src/dotty/tools/dotc/interactive/Completion.scala (hunk 1 above)
Comment (Blaisorblade) [reply compiler/src/dotty/tools/dotc/interactive/Completion.scala]: Your change suggests `addAccessibleMembers` is buggy on `NullType` and `Nothing`, so why doesn't the fix should go closer to the source of the bug? At least in `accessibleMembers`, one would think.
And while `defn.NothingType` probably needs to be a special case, `defn.NullType` inherits from `AnyRefType`, so should inherit its members, no?

File: compiler/src/dotty/tools/dotc/interactive/Completion.scala (hunk 4 above)
Comment (nicolasstucki) [reply compiler/src/dotty/tools/dotc/interactive/Completion.scala]: Done

File: compiler/src/dotty/tools/dotc/interactive/Completion.scala (hunk 5 above)
Comment (nicolasstucki) [reply compiler/src/dotty/tools/dotc/interactive/Completion.scala]: Removed this hack.

File: compiler/src/dotty/tools/dotc/interactive/Completion.scala (hunk 3 above)
Comment (nicolasstucki) [reply compiler/src/dotty/tools/dotc/interactive/Completion.scala]: Removed this hack

File: compiler/src/dotty/tools/dotc/interactive/Completion.scala (hunk 4 above)
Comment (smarter) [reply compiler/src/dotty/tools/dotc/interactive/Completion.scala]: There's `isBottomType`

File: compiler/src/dotty/tools/dotc/interactive/Completion.scala (hunk 1 above)
Comment (smarter) [reply compiler/src/dotty/tools/dotc/interactive/Completion.scala]: Yes, I still don't understand what most of this patch is doing.

Comment (nicolasstucki) [reply compiler/src/dotty/tools/dotc/interactive/Completion.scala]: I refactored the code. Now I only special case implicit completions on `Null`. The description of the PR was updated. The source of the issue where the implicit conversion, not the `addAccessibleMembers`

File: compiler/src/dotty/tools/dotc/interactive/Completion.scala (hunk 4 above)
Comment (nicolasstucki) [reply compiler/src/dotty/tools/dotc/interactive/Completion.scala]: Used `isBottomType`

File: compiler/src/dotty/tools/dotc/interactive/Completion.scala [hunk 6]
```diff
@@ -207,11 +210,13 @@ object Completion {
...
     def addMemberCompletions(qual: Tree)(implicit ctx: Context): Unit = {
-      addAccessibleMembers(qual.tpe)
-      if (!mode.is(Mode.Import)) {
-        // Implicit conversions do not kick in when importing
-        implicitConversionTargets(qual)(ctx.fresh.setExploreTyperState())
-          .foreach(addAccessibleMembers)
+      if (!qual.tpe.widenDealias.isBottomType) {
+        addAccessibleMembers(qual.tpe)
+        if (!mode.is(Mode.Import) && !qual.tpe.isRef(defn.NullClass)) {
+          // Implicit conversions do not kick in when importing
```
Comment (Blaisorblade) [inline compiler/src/dotty/tools/dotc/interactive/Completion.scala]: Ah, this fix makes more sense! 

I wonder why *this* part doesn't work correctly, but looks okay... I'd probably add
```
// and for `NullClass` they produce unapplicable completions (for unclear reasons).
```

File: compiler/src/dotty/tools/dotc/util/NameTransformer.scala [hunk 7]
```diff
@@ -55,6 +55,29 @@ object NameTransformer {
...
+        }
+      }
+      while (i < name.length) {
+        sb.append(name(i))
+        i += 1
+      }
+      sb.result()
+    }
+    else name
+  }
```
Comment (Blaisorblade) [inline compiler/src/dotty/tools/dotc/util/NameTransformer.scala]: Is `decodeIllegalChars` meant to invert `avoidIllegalChars`? Ideally, a few tests `decodeIllegalChars(avoidIllegalChars(str)) == str` would be good.

Potential, shorter alternative for the 2 while loops:
```scala
      while (i < name.length) {
        if (i < name.length - 5 && name(i) == '$' && name(i + 1) == 'u') {
          sb.append(Integer.valueOf(name.substring(i + 2, i + 6), 16).toChar)
          i += 6
        } else {
          sb.append(name(i))
          i += 1
        }
      }
```

**Pull request merged**